README.md
# MALDIquantTools

[![Project Status: Abandoned – Initial development has started, but there has not yet been a stable, usable release; the project has been abandoned and the author(s) do not intend on continuing development.](https://www.repostatus.org/badges/latest/abandoned.svg)](https://www.repostatus.org/#abandoned)

## Description

Additional tools for [MALDIquant](http://strimmerlab.org/software/maldiquant/)
This package is **not** intended for
public release on [CRAN](http://cran.r-project.org/).

This package is abandoned. There is no further development and no guarantee
that it is still working.

## Install

```R
install.packages("devtools")
library("devtools")
install_github("sgibb/MALDIquantTools")
```

## Features

- Calculate FWHM.
- Find reference peaks in each individual spectrum. 
- Find monoisotopic peaks in spectra ([moved to MALDIquant](https://github.com/sgibb/MALDIquant/))
- Draw gelmaps

## Gelmap Example
```r
library("MALDIquant")
library("MALDIquantTools")
data(fiedler2009subset)

## select some tumor/control subset
spectra <- fiedler2009subset[9:16]

## run preprocessing
spectra <- transformIntensity(spectra, method="sqrt")
spectra <- smoothIntensity(spectra, method="MovingAverage", halfWindowSize=2)
spectra <- removeBaseline(spectra, method="SNIP")
spectra <- calibrateIntensity(spectra, method="TIC")
peaks <- detectPeaks(spectra)
peaks <- binPeaks(peaks)

## create labels
tumor <- sapply(peaks, function(x)grepl(pattern="tumor", x=metaData(x)$file))
fullName <- sapply(peaks, function(x)metaData(x)$fullName)
rowLabels <- paste(fullName, " (", ifelse(tumor, "T", "C"), ")", sep="")

## run clustering and create dendrogram
iM <- intensityMatrix(peaks, spectra)
d <- dist(iM, method="euclidean")
d <- as.dendrogram(hclust(d, method="complete"), hang=-1)

## plot gelmap
gelmap(iM, rowLabels=rowLabels, dendrogram=d, xlab="mass [Da]")
```
![gelmap](https://github.com/sgibb/MALDIquantTools/raw/master/images/gelmap.png)


NEWS
RELEASE HISTORY OF THE "MALDIquantTools" PACKAGE
================================================

Version 0.19 [2019-10-16]:
- Adapt to MALDIquant 1.19; Package abandoned.

Version 0.18 [2013-03-02]:
- Some internal changes, now depends on MALDIquant 1.6.

Version 0.17 [2012-12-11]:
- Remove all import/export functions (moved to MALDIquantForeign >= 0.1 on
  CRAN).

Version 0.16 [2012-12-01]:
- monoisotopic: remove "false-positive-rule".
- gelmap: fix x axis labels and dendrogram marigns; add cex.axis argument.

Version 0.15 [2012-11-15]:
- Remove movingAverage (moved to MALDIquant >= 1.4.3).

Version 0.14.1 [2012-11-18]:
- monoisotopic: adjust score calculation.

Version 0.14 [2012-11-08]:
- Rewrite monoisotopic.
- Remove detectMonoisotopicPeaks (it is too ugly, too much arguments, ...
  use detectPeaks and monoisotopic instead).

Version 0.13 [2012-10-18]:
- Add fwhm.

Version 0.12 [2012-10-14]:
- Adapt to snr feature in MALDIquant 1.3.6.
- Remove whichClosest (use MALDIquant:::.which.closest instead; monoistopic peak
  detection is no 2 times faster).
- Remove estimateNoiseSuperSmoother (moved to MALDIquant).

Version 0.11 [2012-10-10]:
- Add detectMonoisotopicPeaks.

Version 0.10 [2012-10-07]:
- Add estimateNoiseSuperSmoother.

Version 0.9.1 [2012-09-19]:
- Add convertFiles (combines readBrukerFlexData/readMzXmlData and export*
  methods).

Version 0.9 [2012-09-09]:
- Add gelmap.

Version 0.8 [2012-08-29]:
- Add monoisotopic peak detection.
- Remove binaryMatrix and isBinaryMatrix
  (now in package similarity: https://github.com/sgibb/similarity).

Version 0.7 [2012-08-14]:
- Add individualReferencePeaks.

Version 0.6 [2012-08-14]:
- Add binningCosts.

Version 0.5.1 [2012-07-25]:
- exportMsd: Fix argument peaks.
- .writeMsdDocument: Fix file.access test.

Version 0.5 [2012-07-25]
- Add binaryMatrix.

Version 0.4 [2012-07-25]:
- Add export to mMass' MSD files.

Version 0.3 [2012-07-23]:
- Add movingAverage.

Version 0.2 [2012-07-22]:
- Add export to ASCII and CSV files.

Version 0.1 [2012-07-19]:
- First public release.
- Import of ASCII, CSV and mzR supported files.



R/gelmap-functions.R
## Copyright 2012 Sebastian Gibb
## <mail@sebastiangibb.de>
##
## This is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## It is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## See <http://www.gnu.org/licenses/>

#' This function draws a simple pseudo-gelmap.
#'
#' @param x an `intensityMatrix`.
#' @param col gelmap colours.
#' @param rowLabels labels of rows.
#' @param dendrogram a \code{\link[stats]{dendrogram}} which is plotted on the
#'  left of the gelmap.
#' @param dendrogramRatio ratio of width of dendrogram to gelmap.
#' @param xlab a label for the x axis.
#' @param cex.axis magnification to be used for axis annotation (relative to
#'  current setting of \code{cex}).
#' @param \ldots further arguments passed to \code{\link[graphics]{image}}.
#'
#' @export
#' @seealso \code{\link[MALDIquant]{intensityMatrix}}
#' @rdname gelmap
#' @examples
#' library("MALDIquant")
#' library("MALDIquantTools")
#' data(fiedler2009subset)
#'
#' ## select some tumor/control subset
#' spectra <- fiedler2009subset[9:16]
#'
#' ## run preprocessing
#' spectra <- transformIntensity(spectra, method="sqrt")
#' spectra <- smoothIntensity(spectra, method="MovingAverage", halfWindowSize=2)
#' spectra <- removeBaseline(spectra, method="SNIP")
#' spectra <- calibrateIntensity(spectra, method="TIC")
#' peaks <- detectPeaks(spectra)
#' peaks <- binPeaks(peaks)
#'
#' ## create labels
#' tumor <- sapply(peaks, function(x)grepl(pattern="tumor", x=metaData(x)$file))
#' fullName <- sapply(peaks, function(x)metaData(x)$fullName)
#' rowLabels <- paste(fullName, " (", ifelse(tumor, "T", "C"), ")", sep="")
#'
#' ## run clustering and create dendrogram
#' iM <- intensityMatrix(peaks, spectra)
#' d <- dist(iM, method="euclidean")
#' d <- as.dendrogram(hclust(d, method="complete"), hang=-1)
#'
#' ## plot gelmap
#' gelmap(iM, rowLabels=rowLabels, dendrogram=d, xlab="mass [Da]")
#'
gelmap <- function(x, col=gray((255:1)/255), rowLabels,
                   dendrogram, dendrogramRatio=1/5, xlab="mass", cex.axis=0.75,
                   ...) {
  ## handle arguments
  optArgs <- list(...)

  stopifnot(is.matrix(x))

  ## backup graphical settings and restore them later
  parSettings <- par(no.readonly=TRUE)
  par(cex.axis=cex.axis)
  par(cex.lab=cex.axis)
  on.exit(par(parSettings))

  nr <- nrow(x)
  nc <- ncol(x)

  idx <- 1:nr

  ## plot dendrogram
  if (!missing(dendrogram)) {
    layout(t(as.matrix(1:2)), widths=c(1, 1/dendrogramRatio),
           heights=rep(1, 2))
    par(mai=c(0.85, 0.25, 0.65, 0))

    if (!inherits(dendrogram, "dendrogram")) {
      dendrogram <- as.dendrogram(dendrogram, hang=-1)
    }

    if (!missing(rowLabels)) {
      newOrder <- sort(rowLabels, index.return=TRUE, decreasing=TRUE)$ix
      dendrogram <- reorder(dendrogram, newOrder)
    }

    idx <- order.dendrogram(dendrogram)
    plot(dendrogram, axes=FALSE, ann=FALSE, horiz=TRUE, xaxs="i",
         leaflab="none", ylim=c(1, nr))
  }

  x <- x[idx, ]

  ## plot gelmap
  leftMargin <- ifelse(missing(dendrogram), 0.25, 0)

  if (!missing(rowLabels)) {
    maxLabelWidth <- max(sapply(rowLabels, strwidth, units="inch"))
    par(mai=c(1, leftMargin, 0.8, 0.25+maxLabelWidth))
  } else {
    par(mai=c(1, leftMargin, 0.8, 0))
  }

  arguments <- list(x=1:nc, y=1:nr, z=t(x), col=col, ylab="", xlab=xlab,
                    axes=FALSE, add=FALSE)
  arguments <- c(arguments, optArgs)

  do.call(image, arguments)

  ## plot y axis
  if (!missing(rowLabels)) {
    rowLabels <- rowLabels[idx]
    axis(4, at=axTicks(4, axp=c(1, nr, nr-1)), labels=rowLabels, tick=FALSE,
         las=1)
  }

  ## plot x axis
  mass <- as.double(colnames(x))
  colLabels <- as.character(round(mass))[seq(1, nc, length.out=10)]
  axis(1, at=axTicks(1, axp=c(1, nc, length(colLabels)-1)), labels=colLabels, las=2)

  invisible(NULL)
}

